### PR TITLE
Add analyzer for weighing-machine

### PR DESCRIPTION
**Pull request by Grenkin1988**
Fixes: https://github.com/exercism/csharp-analyzer/issues/107

### CODE REVIEW

**Comment by Grenkin1988** [comment]
This is the start, but would be nice to hear feedback if I am doing this right

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        protected override SolutionAnalysis DisapproveWhenInvalid(WeighingMachineSolution solution)
+        {
+            if (solution.MissingWeighingMachineClass)
+                solution.AddComment(MissingClass(WeighingMachineClassName));
+
+            foreach (var missing in solution.MissingRequiredProperties())
+            {
+                solution.AddComment(MissingProperty(missing));
+            }
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineAnalyzer.cs:19]: I'm not sure we need this check as the tests won't even compile if a property is missing, but it also doesn't hurt :)

File: src/Exercism.Analyzers.CSharp/Analyzers/Shared/SharedComments.cs [hunk 2]
```diff
@@ -34,5 +34,8 @@ internal static class SharedComments
 
         public static SolutionComment InvalidMethodSignature(string method, string signature) =>
             new SolutionComment("csharp.general.invalid_method_signature", new SolutionCommentParameter(Name, method), new SolutionCommentParameter(Signature, signature));
+
+        public static SolutionComment MissingProperty(string property) =>
+            new("csharp.general.missing_property", new SolutionCommentParameter(Name, property));
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/Shared/SharedComments.cs:39]: FYI For each key of a property, an entry must be added to the corresponding website-copy directory: https://github.com/exercism/website-copy/tree/main/analyzer-comments/csharp

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+            if (!solution.PrecisionIsAutoProperty)
+            {
+                solution.AddComment(new SolutionComment("not PrecisionIsAutoProperty"));
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineAnalyzer.cs:23]: You'll probably want to create a public static member in the `WeighingMachineComments` class for this comment. Same with the comments below.

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSyntaxFactory.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using Exercism.Analyzers.CSharp.Syntax;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Exercism.Analyzers.CSharp.Analyzers.WeighingMachine
+{
+    public static class WeighingMachineSyntaxFactory
+    {
+        public static MemberAccessExpressionSyntax GetMethodCalled(this SyntaxNode syntaxNode, string methodName) =>
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSyntaxFactory.cs:12]: This looks like it could be useful for more than just this analyzer. Maybe `SyntaxNodeExtensions` is a better location?

File: src/Exercism.Analyzers.CSharp/Syntax/PropertyDeclarationSyntaxExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Exercism.Analyzers.CSharp.Syntax
+{
+    public static class PropertyDeclarationSyntaxExtensions
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Syntax/PropertyDeclarationSyntaxExtensions.cs:11]: I like this a lot!

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                (setAccessor.Modifiers.Count == 0 || 
+                !setAccessor.Modifiers.Any(m => m.IsKind(Microsoft.CodeAnalysis.CSharp.SyntaxKind.PrivateKeyword)));
+        }
+
+        public bool WeightFieldNameIsPrivate => WeighingMachineClass?.GetField(WeightFieldName)?.IsPrivate() ?? false;
+
+        public bool IsRoundMethodCalledInDisplayWeightProperty() => 
+            DisplayWeightProperty?.GetMethodCalled("Round") is not null;
+
+
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:66]:
```suggestion        
```

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+namespace Exercism.Analyzers.CSharp.Analyzers.WeighingMachine
+{
+    internal class WeighingMachineSolution : Solution
+    {
+        public const string WeighingMachineClassName = "WeighingMachine";
+        public const string PrecisionPropertyName = "Precision";
+        public const string WeightPropertyName = "Weight";
+        public const string TareAdjustmentPropertyName = "TareAdjustment";
+        public const string DisplayWeightPropertyName = "DisplayWeight";
+        public const string WeightFieldName = "weight";
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:21]: Is this checking the backing field of the property? This could be problematic, as we usually don't comment on the naming of things (unless it breaks the tests) and I'm wondering whether this would still work then?

**Review by ErikSchierboom — COMMENTED** [review]
This is great! I've left some small comments.

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs (hunk 7 above)
Comment (Grenkin1988) [reply src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:21]: The idea was to check that backing field is private. I can try thinking of another approach for this.

**Comment by Grenkin1988** [comment]
Oh, now we need to use VS2020

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs (hunk 7 above)
Comment (ErikSchierboom) [reply src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:19]: Maybe you could try to get the name of the backing field from the property definition?

File: src/Exercism.Analyzers.CSharp/Properties/launchSettings.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Properties/launchSettings.json:1]: I think we should git-ignore these files, right?

**Comment by ErikSchierboom** [comment]
> Oh, now we need to use VS2022

Sure. I'm using Rider BTW.

File: src/Exercism.Analyzers.CSharp/Properties/launchSettings.json (hunk 8 above)
Comment (Grenkin1988) [reply src/Exercism.Analyzers.CSharp/Properties/launchSettings.json:1]: Yes, I think so. I'll revert it.

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs (hunk 7 above)
Comment (Grenkin1988) [reply src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:19]: Yes, I was wondering about this

File: src/Exercism.Analyzers.CSharp/Properties/launchSettings.json (hunk 8 above)
Comment (ErikSchierboom) [reply src/Exercism.Analyzers.CSharp/Properties/launchSettings.json:1]: 👍

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs (hunk 7 above)
Comment (ErikSchierboom) [reply src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineSolution.cs:19]: It will be slightly finicky, but it is doable I think

**Comment by Grenkin1988** [comment]
Comments: https://github.com/exercism/website-copy/pull/2134

**Comment by ErikSchierboom** [comment]
@Grenkin1988 Great work! Before I'll start reviewing, there's one thing missing: tests. With an analyzer, it is really important to have a big set of tests, as we don't want to give the wrong suggestions. See https://github.com/exercism/csharp-analyzer/tree/main/test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/Gigasecond for an example of what a test suite for an analyzer looks like.

**Comment by Grenkin1988** [comment]
> @Grenkin1988 Great work! Before I'll start reviewing, there's one thing missing: tests. With an analyzer, it is really important to have a big set of tests, as we don't want to give the wrong suggestions. See https://github.com/exercism/csharp-analyzer/tree/main/test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/Gigasecond for an example of what a test suite for an analyzer looks like.

Yes, thanks, that was the next step

**Comment by ErikSchierboom** [comment]
Lovely!

**Comment by ErikSchierboom** [comment]
@Grenkin1988 I'll work on a fix in a separate PR. 

N.B. I added https://github.com/exercism/csharp-analyzer/pull/109/commits/f8772dece901cf17dd5f206772e388fc44449873 to this PR

**Comment by ErikSchierboom** [comment]
@Grenkin1988 If you rebase on [this branch](https://github.com/exercism/csharp-analyzer/pull/115), you should not have the error.

**Comment by Grenkin1988** [comment]
@ErikSchierboom merged everything. Looks it is ready for review

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿
+using static Exercism.Analyzers.CSharp.Analyzers.Shared.SharedCommentParameters;
+
+namespace Exercism.Analyzers.CSharp.Analyzers.WeighingMachine
+{
+    internal class WeighingMachineComments
+    {
+        public static SolutionComment PropertyIsNotAutoProperty(string name) =>
+            new("csharp.weighingmachine.property_is_not_auto_property", new SolutionCommentParameter(Name, name));
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs:9]: I think this comment could actually also be useful for other exercises. As such, could you make this be a general comment? This means moving it to the `SharedComments.cs` class and having it start with `csharp.general`.

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+        public static SolutionComment PrecisionPropertyHasNonPrivateSetter(string name) =>
+            new("csharp.weighingmachine.property_setter_is_not_private", new SolutionCommentParameter(Name, name));
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs:12]: I think this comment could actually also be useful for other exercises. As such, could you make this be a general comment? This means moving it to the `SharedComments.cs` class and having it start with `csharp.general`.

File: src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+        public static SolutionComment RoundMethodNotCalledInDisplayWeightProperty =
+            new("csharp.weighingmachine.round_called_in_display_weight");
+
+        public static SolutionComment PropertyBetterUseInitializer(string name) =>
+            new("csharp.weighingmachine.property_better_use_initializer", new SolutionCommentParameter(Name, name));
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Analyzers/WeighingMachine/WeighingMachineComments.cs:18]: I think this comment could actually also be useful for other exercises. As such, could you make this be a general comment? This means moving it to the `SharedComments.cs` class and having it start with `csharp.general`.

File: src/Exercism.Analyzers.CSharp/Exercism.Analyzers.CSharp.csproj [hunk 12]
```diff
@@ -15,5 +15,9 @@
         <PackageReference Include="Serilog" Version="2.10.0" />
         <PackageReference Include="Serilog.Sinks.Console" Version="4.0.1" />
     </ItemGroup>
+    
+    <ItemGroup>
+      <Folder Include="Properties\" />
+    </ItemGroup>
```
Comment (ErikSchierboom) [inline src/Exercism.Analyzers.CSharp/Exercism.Analyzers.CSharp.csproj:21]: Was this needed?

**Review by ErikSchierboom — COMMENTED** [review]
Great work! I've left some minor comments, but it looks impressive! 

I noticed that there was no "approved" test case. Could you add one? That way we know that there _is_ a solution that has no comments. Ideally, this should be the exemplar solution, so maybe you could add that as a test case?

**Comment by Grenkin1988** [comment]
> Great work! I've left some minor comments, but it looks impressive!
> 
> I noticed that there was no "approved" test case. Could you add one? That way we know that there _is_ a solution that has no comments. Ideally, this should be the exemplar solution, so maybe you could add that as a test case?

Done

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj [hunk 13]
```diff
@@ -14,9 +14,19 @@
   <ItemGroup>
     <SolutionFiles Include="Solutions\**" />
     <Compile Remove="@(SolutionFiles)" />
+    <Compile Remove="Solutions\WeighingMachine\Approve\WeighingMachine.cs" />
```
Comment (ErikSchierboom) [inline test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:17]: Could you revert the changes in this file, as I don't think they're needed.

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public int Precision
+    {
+        get { return _precision; }
+        private set
+        {
+            _precision = value;
+        }
+    }
```
Comment (ErikSchierboom) [inline test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs:20]: I think we'd want people to do:

```suggestion
    public int Precision { get; private set; }
```

It might be good to add a rule for that in the analyzer too.

**Review by ErikSchierboom — COMMENTED** [review]
Lovely. Just some minor comments!

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs (hunk 14 above)
Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs]: Rule already exist, looks like test wasn't running before

Comment (ErikSchierboom) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs]: Do we have a test case that verifies that a comment is made when the student uses:

```csharp
public int Precision
    {
        get { return _precision; }
        private set
        {
            _precision = value;
        }
    }
```
?

Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs]: It checks that autoproperty is used

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj [hunk 15]
```diff
@@ -14,6 +14,9 @@
   <ItemGroup>
     <SolutionFiles Include="Solutions\**" />
     <Compile Remove="@(SolutionFiles)" />
+    <Content Include="@(SolutionFiles)">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
```
Comment (ErikSchierboom) [inline test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:19]: I think this can be removed.

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs (hunk 14 above)
Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions/WeighingMachine/Approve/WeighingMachine.cs]: https://github.com/exercism/csharp-analyzer/pull/109/files#diff-91639737ce24cbf45cff1d2e6fed7e8ca2ef944bdc33cb5e2549bcb85f3786e2R24

File: test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj (hunk 15 above)
Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:19]: If this is removed, Cs files are not in VS

Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:19]: image

Comment (Grenkin1988) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:19]: image

Comment (ErikSchierboom) [reply test/Exercism.Analyzers.CSharp.IntegrationTests/Exercism.Analyzers.CSharp.IntegrationTests.csproj:19]: Aha. Rider behaves differently then. Okay cool.

**Comment by ErikSchierboom** [comment]
@Grenkin1988 Is https://github.com/exercism/website-copy/pull/2134 up to date with the latest renames?

**Comment by Grenkin1988** [comment]
> @Grenkin1988 Is https://github.com/exercism/website-copy/pull/2134 up to date with the latest renames? 

I will check, but it should be

**Comment by ErikSchierboom** [comment]
Okay. I'll do a check on the text.

**Comment by ErikSchierboom** [comment]
I've reviewed the website-copy PR and left some textual comments.

Great work on getting this PR merged! If you found this fun, there are tons more analyzers to implement :)
